File: Platform/infra/http/interfaces/dotnet/HttpClient.cs
using System.Text.Json;

namespace Coyote.Infra.Http;

/// <summary>
/// HTTP method enumeration
/// </summary>
public enum HttpMethod
{
    Get,
    Post,
    Put,
    Delete,
    Patch,
    Head,
    Options
}

/// <summary>
/// HTTP response contract
/// </summary>
public interface IHttpResponse
{
    /// <summary>
    /// HTTP status code
    /// </summary>
    int StatusCode { get; }
    
    /// <summary>
    /// Response body as string
    /// </summary>
    string Body { get; }
    
    /// <summary>
    /// Response headers
    /// </summary>
    IReadOnlyDictionary<string, string> Headers { get; }
    
    /// <summary>
    /// Indicates if the response represents a successful HTTP status (2xx)
    /// </summary>
    bool IsSuccess { get; }
    
    /// <summary>
    /// Error message if the request failed
    /// </summary>
    string? ErrorMessage { get; }
    
    /// <summary>
    /// Get header value by name (case-insensitive)
    /// </summary>
    string? GetHeader(string name);
    
    /// <summary>
    /// Deserialize response body to specified type
    /// </summary>
    T? GetContent<T>();
}

/// <summary>
/// HTTP request configuration
/// </summary>
public interface IHttpRequest
{
    /// <summary>
    /// Request URL
    /// </summary>
    string Url { get; set; }
    
    /// <summary>
    /// HTTP method
    /// </summary>
    HttpMethod Method { get; set; }
    
    /// <summary>
    /// Request body
    /// </summary>
    string? Body { get; set; }
    
    /// <summary>
    /// Request headers
    /// </summary>
    IDictionary<string, string> Headers { get; }
    
    /// <summary>
    /// Request timeout in milliseconds
    /// </summary>
    int? TimeoutMs { get; set; }
    
    /// <summary>
    /// Client certificate path
    /// </summary>
    string? ClientCertPath { get; set; }
    
    /// <summary>
    /// Client certificate key path
    /// </summary>
    string? ClientKeyPath { get; set; }
    
    /// <summary>
    /// CA certificate path for server verification
    /// </summary>
    string? CACertPath { get; set; }
    
    /// <summary>
    /// Whether to verify server certificate
    /// </summary>
    bool VerifyPeer { get; set; }
    
    /// <summary>
    /// Whether to follow redirects
    /// </summary>
    bool FollowRedirects { get; set; }
    
    /// <summary>
    /// Set request body as JSON
    /// </summary>
    void SetJsonBody<T>(T content);
    
    /// <summary>
    /// Add or update header
    /// </summary>
    void SetHeader(string name, string value);
}

/// <summary>
/// Main HTTP client interface for the Coyote platform
/// </summary>
public interface ICoyoteHttpClient : IDisposable
{
    /// <summary>
    /// Execute a configured HTTP request
    /// </summary>
    Task<IHttpResponse> ExecuteAsync(IHttpRequest request, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Perform GET request
    /// </summary>
    Task<IHttpResponse> GetAsync(string url, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Perform POST request
    /// </summary>
    Task<IHttpResponse> PostAsync(string url, string? body = null, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Perform POST request with JSON body
    /// </summary>
    Task<IHttpResponse> PostJsonAsync<T>(string url, T content, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Perform PUT request
    /// </summary>
    Task<IHttpResponse> PutAsync(string url, string? body = null, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Perform PUT request with JSON body
    /// </summary>
    Task<IHttpResponse> PutJsonAsync<T>(string url, T content, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Perform DELETE request
    /// </summary>
    Task<IHttpResponse> DeleteAsync(string url, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Perform PATCH request
    /// </summary>
    Task<IHttpResponse> PatchAsync(string url, string? body = null, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Set default timeout for all requests
    /// </summary>
    void SetDefaultTimeout(int timeoutMs);
    
    /// <summary>
    /// Set default headers for all requests
    /// </summary>
    void SetDefaultHeaders(IReadOnlyDictionary<string, string> headers);
    
    /// <summary>
    /// Set client certificate for mTLS
    /// </summary>
    void SetClientCertificate(string certPath, string keyPath);
    
    /// <summary>
    /// Set CA certificate for server verification
    /// </summary>
    void SetCACertificate(string caPath);
    
    /// <summary>
    /// Set whether to verify server certificates
    /// </summary>
    void SetVerifyPeer(bool verify);
    
    /// <summary>
    /// Test connectivity to specified URL
    /// </summary>
    Task<bool> PingAsync(string url, CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Create a new request builder
    /// </summary>
    IHttpRequest CreateRequest();
}

File: Platform/infra/http/src/python/interfaces/http_client.py
"""
HTTP Client Interface Module

This module provides abstract base classes for HTTP client functionality,
mirroring the C++ interface architecture but using Python conventions.
"""

from abc import ABC, abstractmethod
from enum import Enum, auto
from typing import Dict, Optional, Any
import dataclasses


class HttpMethod(Enum):
    """HTTP method enumeration."""
    GET = auto()
    POST = auto()
    PUT = auto()
    DELETE = auto()
    PATCH = auto()
    HEAD = auto()
    OPTIONS = auto()


@dataclasses.dataclass
class HttpResponse:
    """HTTP response data container."""
    status_code: int
    body: str
    headers: Dict[str, str]
    error_message: str = ""
    
    @property
    def is_success(self) -> bool:
        """Check if the response indicates success (2xx status code)."""
        return 200 <= self.status_code < 300
    
    def get_header(self, name: str, default: Optional[str] = None) -> Optional[str]:
        """Get a header value by name (case-insensitive)."""
        for key, value in self.headers.items():
            if key.lower() == name.lower():
                return value
        return default


@dataclasses.dataclass
class HttpRequest:
    """HTTP request data container."""
    url: str = ""
    method: HttpMethod = HttpMethod.GET
    body: str = ""
    headers: Dict[str, str] = dataclasses.field(default_factory=dict)
    timeout_ms: int = 10000
    client_cert_path: str = ""
    client_key_path: str = ""
    ca_cert_path: str = ""
    verify_peer: bool = True
    follow_redirects: bool = True
    
    def set_header(self, key: str, value: str) -> None:
        """Set a header value."""
        self.headers[key] = value
    
    def update_headers(self, headers: Dict[str, str]) -> None:
        """Update multiple headers."""
        self.headers.update(headers)


class HttpClient(ABC):
    """Abstract base class for HTTP clients."""
    
    @abstractmethod
    def execute(self, request: HttpRequest) -> HttpResponse:
        """Execute an HTTP request and return the response."""
        pass
    
    @abstractmethod
    def get(self, url: str, headers: Optional[Dict[str, str]] = None) -> HttpResponse:
        """Perform a GET request."""
        pass
    
    @abstractmethod
    def post(self, url: str, body: str = "", headers: Optional[Dict[str, str]] = None) -> HttpResponse:
        """Perform a POST request."""
        pass
    
    @abstractmethod
    def put(self, url: str, body: str = "", headers: Optional[Dict[str, str]] = None) -> HttpResponse:
        """Perform a PUT request."""
        pass
    
    @abstractmethod
    def delete(self, url: str, headers: Optional[Dict[str, str]] = None) -> HttpResponse:
        """Perform a DELETE request."""
        pass
    
    @abstractmethod
    def set_default_timeout(self, timeout_ms: int) -> None:
        """Set the default timeout for all requests."""
        pass
    
    @abstractmethod
    def set_default_headers(self, headers: Dict[str, str]) -> None:
        """Set default headers for all requests."""
        pass
    
    @abstractmethod
    def set_client_certificate(self, cert_path: str, key_path: str) -> None:
        """Set client certificate for SSL authentication."""
        pass
    
    @abstractmethod
    def set_ca_certificate(self, ca_path: str) -> None:
        """Set CA certificate for SSL verification."""
        pass
    
    @abstractmethod
    def set_verify_peer(self, verify: bool) -> None:
        """Enable or disable SSL peer verification."""
        pass
    
    @abstractmethod
    def ping(self, url: str) -> bool:
        """Check connectivity to a URL."""
        pass